Lib/test/test_asyncgen.py

```
@@ -405,6 +406,26 @@ async def test_2():
         result = self.loop.run_until_complete(test_2())
         self.assertEqual(result, "completed")
 
+        def test_send():
+            p = ait_class()
+            obj = anext(p, "completed")
+            try:
+                with contextlib.closing(obj.__await__()) as g:
+                    g.send(None)
+            except StopIteration:
+                pass
+
+        test_send()
+
+        async def test_throw():
+            p = ait_class()
+            obj = anext(p, "completed")
+            self.assertRaises(SyntaxError, obj.throw, SyntaxError)
```

this is another one where `ait_class` would expect to have a valid `current_task()` calling the throw method

---

I think it would be good to get a test of an anext where the asyncgen suppresses the throw:

```python
@types.coroutine
def _async_yield(v):
    return (yield v)

class MyError(Exception):
    pass

async def agenfn():
    try:
        await _async_yield(1)
    except MyError:
        await _async_yield(2)
    return
    yield

agen = agenfn()
with contextlib.closing(anext(agen, "default").__await__()) as g:
    assert g.send(None) == 1
    assert g.throw(MyError, MyError(), None) == 2
    try:
        g.send(None)
    except StopIteration as e:
        err = e
    assert err.value == "default"
```

---

You need to remove the `return` from `agenfn` to actually raise a StopIteration instead of StopAsyncIteration. I also don't understand why call `__await__()` here since that just returns self in case of `anext`'s result.

But yeah, the last line fails now because `value` is `None` instead of `"default"`.

---

The agen should raise StopAsyncIteration which is intercepted by anext and replaced with StopIteration ("default")

---

I think both send() and throw() (and __next__()) need to have the behaviour of converting StopAsyncIteration to StopIteration(default).

---

@lordmauve @graingert Thanks guys, I've pushed an update with many more tests and fixes to the C code. You're right that `StopAsyncIteration` needed to be wrapped to `StopIteration`.